### ISSUE TITLE
How to pass custom metadata during authentication failure

### ISSUE THREAD

**Issue by jayadeep230891** [LogNet/grpc-spring-boot-starter#189]
I have enabled spring security for my project. Currently i see no way to send custom metadata(In my case transaction id for tracking purpose) during an authentication error. I am using a global interceptor onHalfClose method to handle exceptions to send custom metadata as trailers, which i see is not possible in auth errors. While debugging further i saw that in org.lognet.springboot.grpc.security.SecurityInterceptor interceptCall method all authentication exception's are wrapped and the server call is closed with an new instance of metadata object. Because of this either request's metadata is not being reused or there is another way to pass custom headers in the response. 

The only way i see is to override the SecurityInterceptor  by creating a custom @EnableGrpcSecurity, overriden GrpcSecurity and new SecurityInterceptor. 

Is there something I am missing or is this not possible on the current version.

**Comment by jayadeep230891** [comment]
@jvmlet : Could you please have a look into my query posted above.

**Comment by jvmlet** [comment]
Sure, I'll add the callback to enrich metadata

**Comment by jvmlet** [comment]
@jayadeep230891 , is  the (unauthorize/unauthenticated) request message would be enough for you to provide the header values for you ?

**Comment by jayadeep230891** [comment]
@jvmlet sorry, i am a bit lost. Could you explain what you meant by request message ?. Do you mean request body ?

What i was looking for is like "x-request-id: 6d4070b0-9e0d-4193-acdd-c39f34134f28" as metadata. This will help us track all requests including failures(I am already intercepting all requests for tracing and logging them based on the x-request-id).

**Comment by jvmlet** [comment]
Yes, I mean request body. What is your logic to calculate  `6d4070b0-9e0d-4193-acdd-c39f34134f28` value?

**Comment by jayadeep230891** [comment]
I don't think request body is what exactly I am looking for. I feel that its better to avoid trace elements like x-request-id to be passed in body. 

Maybe I will explain our implementation a bit.

During a request arrival, We look for x-request-id in header metadata data. if its available I trace the entire request lifecycle (tracing, logging request response) against the same, else I create one(A ID similar to UUID) and send it as a response metadata. For Exceptions, we send the same as trailer data (Using a custom exception handler interceptor). 

But right now, since the call is getting closed in SecurityInterceptor and my custom exception handler is not Invoked. I was wondering if there is a way to allow metadata to be passed to SecurityInterceptor for your usage in server.close()

As you see the x-request-id (6d4070b0-9e0d-4193-acdd-c39f34134f28) has got nothing to do with request and response body. Its just an extra element to help us trace requests across distributed systems.

**Comment by jvmlet** [comment]
OK. I think I got you. You need the request headers to set the response header.

**Comment by jvmlet** [comment]
I'll also add ordering configuration support for starter's interceptors, so you will be able to place yours at the appropriate place.

**Comment by jayadeep230891** [comment]
Sure that would help. Thank you very much :)

**Comment by jvmlet** [comment]
@jayadeep230891 , please upgrade meanwhile  to `4.4.5-SNAPSHOT`
You can use below properties :
* `grpc.security.auth.interceptor-order` ( defaults to `Ordered.HIGHEST_PRECEDENCE`)
* `grpc.validation.interceptor-order` ( defaults to `Ordered.HIGHEST_PRECEDENCE+10`)
* `grpc.metrics.interceptor-orde`r ( defaults to `Ordered.HIGHEST_PRECEDENCE+20`)

Will still need to support `authFailedCallback` ...

**Comment by jayadeep230891** [comment]
@jvmlet : Sorry for the delay, I will try today and let you know. :)

**Comment by jayadeep230891** [comment]
Hello @jvmlet , I have added snapshot and implemented changes in my current project. The changes itself works fine for the given requirement. But I have observed that the authentication logic in interceptCall from SecurityInterceptor gets executed before onMessage method call of higher priority custom server interceptor.  This is because Logic in intercept call is getting executed before OnMessage method.

While this is not a wrong thing. The difficulty i am facing is that, I wrote the logic to log audit and payload as part of customInterceptor in onMessage Method. But the GRPCErrorHandler.handle kicks in first and missed few details i enrich as part of OnMessage method. 

**Current flow:**
CustomInterceptor(interceptCall ) --> SecurityInterceptor(interceptCall) --> CustomInterceptor(OnMessage) --> CustomInterceptor(onComplete)

**Needed Flow: **
CustomInterceptor(interceptCall,OnMessage)  -->  SecurityInterceptor(interceptCall) --> CustomInterceptor(onComplete) 


My Only option is to skip Incoming log for authentication errors and use request header to enrich the required details for auth failures as a separate flow .  Do you think there is a better way to handle this ?

**Comment by jvmlet** [comment]
Try setting `grpc.security.auth.fail-fast` property to false

**Comment by jayadeep230891** [comment]
Hello @jvmlet : I have tried the suggested property and the execution order is fine. But somewhere internally in onHalfClose method of servercalls.java, the call is getting closed again causing this exception. Please find the stack trace for the same.

java.lang.IllegalStateException: call already closed\r\n\tat com.google.common.base.Preconditions.checkState(Preconditions.java:508)\r\n\tat io.grpc.internal.ServerCallImpl.closeInternal(ServerCallImpl.java:209)\r\n\tat io.grpc.internal.ServerCallImpl.close(ServerCallImpl.java:202)\r\n\tat io.grpc.PartialForwardingServerCall.close(PartialForwardingServerCall.java:48)\r\n\tat io.grpc.ForwardingServerCall.close(ForwardingServerCall.java:22)\r\n\tat io.grpc.ForwardingServerCall$SimpleForwardingServerCall.close(ForwardingServerCall.java:39)\r\n\tat io.grpc.PartialForwardingServerCall.close(PartialForwardingServerCall.java:48)\r\n\tat io.grpc.ForwardingServerCall.close(ForwardingServerCall.java:22)\r\n\tat io.grpc.ForwardingServerCall$SimpleForwardingServerCall.close(ForwardingServerCall.java:39)\r\n\tat org.lognet.springboot.grpc.autoconfigure.metrics.GRpcMetricsAutoConfiguration$MonitoringServerCall.close(GRpcMetricsAutoConfiguration.java:96)\r\n\tat io.grpc.PartialForwardingServerCall.close(PartialForwardingServerCall.java:48)\r\n\tat io.grpc.ForwardingServerCall.close(ForwardingServerCall.java:22)\r\n\tat io.grpc.ForwardingServerCall$SimpleForwardingServerCall.close(ForwardingServerCall.java:39)\r\n\tat io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:176)\r\n\tat io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:35)\r\n\tat io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23)\r\n\tat io.grpc.ForwardingServerCallListener$SimpleForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:40)\r\n\tat io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:35)\r\n\tat io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23)\r\n\tat io.grpc.ForwardingServerCallListener$SimpleForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:40)\r\n\tat io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:35)\r\n\tat io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23)\r\n\tat io.grpc.ForwardingServerCallListener$SimpleForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:40)\r\n\tat io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:35)\r\n\tat io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23)\r\n\tat io.grpc.ForwardingServerCallListener$SimpleForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:40)\r\n\tat com.ingka.spe.grpc.audit.GrpcLoggingInterceptor$2.onHalfClose(GrpcLoggingInterceptor.java:106)\r\n\tat io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:331)\r\n\tat io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:797)\r\n\tat io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)\r\n\tat io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)\r\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\r\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\r\n\tat java.base/java.lang.Thread.run(Thread.java:834)\r\n

Could you please have a look.

**Comment by jvmlet** [comment]
Yes, I saw this in my tests. Not sure why `close` is called twice. BTW, in all previous versions of the starter the behavior was the same and I observed the 
` call already closed` exception when closing the call manually. If it doesn't affect your application logic I would leave it as is. I'll try to get more info from grpc team to understand how to get rid of this exception

**Comment by jayadeep230891** [comment]
@jvmlet : I was not able to debug fully, since there seems to be some issue with debug from 4.4.4 version(Intellij keeps complaining that byte code doesn't match the source code). But what i observed is that the request object in onHalfClose() of ServerCalls is null. This is where they are manually closing the call again(line no 176). I am not sure if this is of much help, but wanted to share.

But yes i understand, I will leave the behavior as is for now. Thank you very much for your help :)
